components/layout_2020/flow/mod.rs

```diff
@@ -486,23 +541,155 @@ fn layout_in_flow_replaced_block_level<'a>(
     let intrinsic_size = replaced.intrinsic_size.size_to_flow_relative(mode);
     // FIXME(nox): This can divide by zero.
     let intrinsic_ratio = intrinsic_size.inline.px() / intrinsic_size.block.px();
+
     let box_size = style.box_size();
+    let min_box_size = style.min_box_size();
+    let max_box_size = style.max_box_size();
+
     let inline_size = box_size.inline.percentage_relative_to(cbis);
-    let block_size = box_size
+    let min_inline_size = min_box_size
+        .inline
+        .percentage_relative_to(cbis)
+        .auto_is(Length::zero);
+    let max_inline_size = match max_box_size.inline {
+        MaxSize::LengthPercentage(max_inline_size) => {
+            Some(max_inline_size.percentage_relative_to(cbis))
+        },
+        MaxSize::None => None,
+    };
+    let block_size = box_size.block.maybe_percentage_relative_to(cbbs);
+    let min_block_size = min_box_size
         .block
-        .maybe_percentage_relative_to(containing_block.block_size.non_auto());
+        .maybe_percentage_relative_to(cbbs)
+        .auto_is(Length::zero);
+    let max_block_size = match max_box_size.block {
+        MaxSize::LengthPercentage(max_block_size) => {
+            max_block_size.maybe_percentage_relative_to(cbbs)
+        },
+        MaxSize::None => None,
+    };
```

Comment:
Looks like this `match` could be a method of `MaxSize`?

Reply:
I didn't want to because we actually unwrap the `NonNegativeLengthPercentage` in `StyleExt::max_box_size`, and it feels weird to me to add methods to `MaxSize<LengthPercentage>` in stylo if they are only used in layout. Maybe at some point we should consider bringing back `NonNegativeLengthPercentage` and `NonNegativeLength` in the layout code. I feel like this would be helpful avoid cases where we end up with negative lengths too anyway.